File: web/src/api/debates.ts
import type {
  CreateDebateRequest,
  DebateActionPayload,
  DebateDetailPayload,
  DebateExecutionPayload,
  DebateListPayload,
  DebateMessagePayload,
  DebateMessagesPayload,
  DebatePersonalitiesPayload,
  DebateVotesPayload,
  StartDebateRequest,
} from "@/types/debates"
import request from "@/utils/request"

const Api = {
  Root: "/api/debates",
  Personalities: "/api/debates/personalities",
  Detail: "/api/debates/{id}",
  Start: "/api/debates/{id}/start",
  Cancel: "/api/debates/{id}/cancel",
  Execute: "/api/debates/{id}/execute",
  Messages: "/api/debates/{id}/messages",
  Votes: "/api/debates/{id}/votes",
} as const

export function getDebatesApi() {
  return request.get<DebateListPayload>(Api.Root)
}

export function createDebateApi(data: CreateDebateRequest) {
  return request.post<DebateActionPayload>(Api.Root, data)
}

export function getDebatePersonalitiesApi() {
  return request.get<DebatePersonalitiesPayload>(Api.Personalities)
}

export function getDebateApi(id: string) {
  return request.get<DebateDetailPayload>(Api.Detail.replace("{id}", id))
}

export function deleteDebateApi(id: string) {
  return request.delete<DebateMessagePayload>(Api.Detail.replace("{id}", id))
}

export function startDebateApi(id: string, data: StartDebateRequest) {
  return request.post<DebateActionPayload>(Api.Start.replace("{id}", id), data)
}

export function cancelDebateApi(id: string) {
  return request.post<DebateActionPayload>(Api.Cancel.replace("{id}", id))
}

export function executeDebateApi(id: string) {
  return request.post<DebateExecutionPayload>(Api.Execute.replace("{id}", id))
}

export function getDebateMessagesApi(id: string) {
  return request.get<DebateMessagesPayload>(Api.Messages.replace("{id}", id))
}

export function getDebateVotesApi(id: string) {
  return request.get<DebateVotesPayload>(Api.Votes.replace("{id}", id))
}

File: core/src/http/handlers/debates.rs
use axum::{
    Json,
    extract::{Path, State},
};

use crate::{
    contracts::debates::{
        CreateDebateRequest, DebateActionPayload, DebateDetailPayload, DebateExecutionPayload,
        DebateListPayload, DebateMessagePayload, DebateMessagesPayload, DebatePersonalitiesPayload,
        DebateVotesPayload, StartDebateRequest,
    },
    error::Result,
    http::{extractors::AuthUser, response::ApiResponse},
    state,
};

pub async fn handle_get_debates(
    State(app): State<state::AppState>,
    user: AuthUser,
) -> Result<Json<ApiResponse<DebateListPayload>>> {
    let payload = app.services.debate_service.list(&user.sub).await;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_create_debate(
    State(app): State<state::AppState>,
    user: AuthUser,
    Json(request): Json<CreateDebateRequest>,
) -> Result<Json<ApiResponse<DebateActionPayload>>> {
    let payload = app
        .services
        .debate_service
        .create(&user.sub, request)
        .await?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_get_debate_personalities(
    State(app): State<state::AppState>,
) -> Result<Json<ApiResponse<DebatePersonalitiesPayload>>> {
    let payload = app.services.debate_service.personalities();
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_get_debate(
    State(app): State<state::AppState>,
    user: AuthUser,
    Path(id): Path<String>,
) -> Result<Json<ApiResponse<DebateDetailPayload>>> {
    let payload = app.services.debate_service.get(&user.sub, &id).await?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_delete_debate(
    State(app): State<state::AppState>,
    user: AuthUser,
    Path(id): Path<String>,
) -> Result<Json<ApiResponse<DebateMessagePayload>>> {
    let payload = app.services.debate_service.delete(&user.sub, &id).await?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_start_debate(
    State(app): State<state::AppState>,
    user: AuthUser,
    Path(id): Path<String>,
    Json(_request): Json<StartDebateRequest>,
) -> Result<Json<ApiResponse<DebateActionPayload>>> {
    let payload = app.services.debate_service.start(&user.sub, &id).await?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_cancel_debate(
    State(app): State<state::AppState>,
    user: AuthUser,
    Path(id): Path<String>,
) -> Result<Json<ApiResponse<DebateActionPayload>>> {
    let payload = app.services.debate_service.cancel(&user.sub, &id)?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_execute_debate(
    State(app): State<state::AppState>,
    user: AuthUser,
    Path(id): Path<String>,
) -> Result<Json<ApiResponse<DebateExecutionPayload>>> {
    let payload = app
        .services
        .debate_service
        .execution(&user.sub, &id)
        .await?;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_get_debate_messages(
    State(app): State<state::AppState>,
    user: AuthUser,
    Path(id): Path<String>,
) -> Result<Json<ApiResponse<DebateMessagesPayload>>> {
    let payload = app.services.debate_service.messages(&user.sub, &id).await;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}

pub async fn handle_get_debate_votes(
    State(app): State<state::AppState>,
    user: AuthUser,
    Path(id): Path<String>,
) -> Result<Json<ApiResponse<DebateVotesPayload>>> {
    let payload = app.services.debate_service.votes(&user.sub, &id).await;
    Ok(Json(ApiResponse::success(Some(payload), None)))
}
